Witness function tests

edrumwri opened the pull request:
A set of more exhaustive tests for witness functions to address issue #6184. Checks that correct behavior is exhibited when:

* Multiple witness functions trigger simultaneously
* Witness function zeros are isolated to the correct tolerance when using fixed step integration
* Witness function zeros are isolated with greater accuracy as it is requested
* Multiple witness functions trigger over a large interval, then fewer witness functions trigger, as the zero(s) are isolated

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6255)
<!-- Reviewable:end -->

edrumwri commented:
+@siyuanfeng-tri for feature review please (I'm hoping reviewing this will give a little context to my updates to the event-API-update PR as well)

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-:-KlzqTZuhxujsAQ6LAAk:b393si4)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
I am actually pretty confused about the FixedStepNoIsolation. Maybe you can help me understand why the witness function fires at 0.001? 

---

Reviewed 2 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/analysis/BUILD, line 165 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km08QwhWdFVTmW6Qq_4:-Km08QwhWdFVTmW6Qq_5:b-v4egvw) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/BUILD#L165)):*
> ```Python
>         ":explicit_euler_integrator",
>         ":implicit_euler_integrator",
>         ":runge_kutta3_integrator",
> ```

should stay below. 

---

*[drake/systems/analysis/test/simulator_test.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km09nVQ6MoQ-SWZ5EkE:-Km09nVQ6MoQ-SWZ5EkF:b-kjrh7l) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L109)):*
> ```cpp
>   LogisticSystem system(1e-8, 100, 1);
>   double publish_time = 0;
>   system.set_publish_callback([&](const Context<double>& context){
> ```

space 

---

*[drake/systems/analysis/test/simulator_test.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0DWl2c8ryYMNDZTVy:-Km0DWl2c8ryYMNDZTVz:b-xzb44x) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L124)):*
> ```cpp
> 
>   // Verify that the witness function triggered at dt.
>   EXPECT_EQ(publish_time, dt);
> ```

why is the witness function fired at dt? I think dx / dt at t = 0 is zero. After 1 tick x should stay at -1? 

---

*[drake/systems/analysis/test/simulator_test.cc, line 128 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0GNg8KblWISHt0lGy:-Km0GNg91DljiuvWZ8hL:b-r1o549) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L128)):*
> ```cpp
> 
> // Tests witness function isolation when operating in error-controlled
> // integration mode.
> ```

Can you add a high level brief description of the behavior here? 

---

*[drake/systems/analysis/test/simulator_test.cc, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0DrlUzWplHUzp_cdP:-Km0DrlUzWplHUzp_cdQ:b-qbqp8g) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L131)):*
> ```cpp
> GTEST_TEST(SimulatorTest, VariableStepIsolation) {
>   // Set empty system to trigger when time is +1.
>   EmptySystem system(+1, WitnessFunction<double>::DirectionType::kCrossesZero);
> ```

BTW, do you really want the +? 

---

*[drake/systems/analysis/test/simulator_test.cc, line 167 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0Gbp6LeSZ2fMGzLAj:-Km0Gbp6LeSZ2fMGzLAk:bd6fgur) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L167)):*
> ```cpp
>     // Simulate to dt.
>     simulator.StepTo(dt);
> ```

BTW, i am not that familiar with the error controlling in the integrator. This system has zero derivative, what are the stopping time in this case? 1 + accuracy? 

---

*[drake/systems/analysis/test/simulator_test.cc, line 184 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0HwK3PaCaWKeT2A7H:-Km0HwK3PaCaWKeT2A7I:b-gcprxs) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L184)):*
> ```cpp
> 
> // Tests witness function isolation accuracy increases with increasing accuracy
> // in the context.
> ```

BTW, Is this test supposed to be functionally equivalent to the one above? the only difference is in the last one, the accuracy is set in the integrator, here it's set in the context? If so, can you add a comment to make it clear? Also consider reduce duplicate code. 

---

*[drake/systems/analysis/test/simulator_test.cc, line 230 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0IGI9gMd2y1Y5Me8E:-Km0IGI9gMd2y1Y5Me8F:b-e5rxji) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L230)):*
> ```cpp
> 
>     // Increase the accuracy.
>     accuracy *= 0.1;
> ```

BTW is there any reason you shrink it slower here than above? 

---

*[drake/systems/analysis/test/simulator_test.cc, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0QHQK8N7gavo88VIp:-Km0QHQK8N7gavo88VIq:bouw4du) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L252)):*
> ```cpp
> 
>     // Evaluate them.
>     double w1 = witnesses.front()->Evaluate(context);
> ```

BTW it'd be nice to say w1 is clock, w2 is logistic witness. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-:-Km0ZqQmOQKl2QMHdFeI:b-rs2r7t)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
checkpoint, i'd like to really understand the first test case/ 

---

Reviewed 1 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/analysis/test/simulator_test.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0DWl2c8ryYMNDZTVy:-Km3LHk7MSeunGvCZjxg:bkhp13l) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L124)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

You're right that the derivative computation at t = 0 is zero. If the Euler integrator were used, x(dt) = -1, as you thought.

However, the derivative computation at t > 0 is no longer zero and using the default RK2 integrator, which "samples" several times in [0, dt], the state at dt will be slightly greater than zero. To better enforce this, I'm guaranteeing that RK2 is used by default.
</blockquote></details>

ah, it's RK 2, fine.
When I got the branch and printed out the time when derivatives are evaled, I only see it at 0 and dt, isn't rk2 supposed to eval at 0.5 dt? I haven't looked at the integrator code. 
In any case, if it's RK2, it will cross zero here, I am wondering why it's returning dt not the time when it crossed 0, is it because of some accuracy setting vs "fixed time mode" you mentioned below? 

---

*[drake/systems/analysis/test/simulator_test.cc, line 131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0DrlUzWplHUzp_cdP:-Km3GBLRA_tef6IOjM_j:b5xphco) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L131)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Changing it to 1.0, which hopefully bothers you less...
</blockquote></details>

I am not really bothered, but some might be.. 

---

*[drake/systems/analysis/test/simulator_test.cc, line 230 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0IGI9gMd2y1Y5Me8E:-Km3KYuS-IQ97rH2-x9q:b-57avt2) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L230)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Are you asking why I tighten the accuracy? I want to ensure that the witness function evaluation gets better as I request greater accuracy.
</blockquote></details>

I mean you are *= 0.1 here instead of *= 0.01, also I am curious is there any absolute numerical statements you can make here. as opposed to new_eval <= eval. 

---

*[drake/systems/analysis/test/simulator_test.cc, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0QHQK8N7gavo88VIp:-Km3LA9TcoNX3CXa00BM:bypeb4b) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L252)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

no change? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-:-Km3LmtdIqwyxioze3q0:b3hx9d8)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/analysis/test/simulator_test.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0DWl2c8ryYMNDZTVy:-Km3MjmHsn4GHMGze45Z:b-za98zs) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L124)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

ah, it's RK 2, fine.
When I got the branch and printed out the time when derivatives are evaled, I only see it at 0 and dt, isn't rk2 supposed to eval at 0.5 dt? I haven't looked at the integrator code. 
In any case, if it's RK2, it will cross zero here, I am wondering why it's returning dt not the time when it crossed 0, is it because of some accuracy setting vs "fixed time mode" you mentioned below? 
</blockquote></details>

Ah, yes, sorry, I should have answered the implicit question. PTAL at the new documentation.

---

*[drake/systems/analysis/test/simulator_test.cc, line 230 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0IGI9gMd2y1Y5Me8E:-Km3QsGYEc40JVeDzNTm:b44d456) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L230)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

I mean you are *= 0.1 here instead of *= 0.01, also I am curious is there any absolute numerical statements you can make here. as opposed to new_eval <= eval. 
</blockquote></details>

Ah, I see. The reason is that it did not shrink sufficiently when using 0.1 (above), where 0.1 was sufficient for a decrease here. No, there's no guarantee that increasing accuracy yields tighter accuracy on the witness function evaluation. Ideally, that would be the case, but the variable step integrator (used for the 0.01 case) means that it's not guaranteed (b/c you don't know the sequence of steps). Updated documentation hopefully explains.

---

*[drake/systems/analysis/test/simulator_test.cc, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0QHQK8N7gavo88VIp:-Km3PDsv3zjADFjmDe0C:b-gubijw) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L252)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

no change? 
</blockquote></details>

I changed `w1` to `clock_eval` and `w2` to `logistic_eval`. What did you want me to change instead?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 2 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/analysis/test/simulator_test.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0DWl2c8ryYMNDZTVy:-Km5qnX3kDSsjwc0Tpbs:brqq2ru) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L124)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Ah, yes, sorry, I should have answered the implicit question. PTAL at the new documentation.
</blockquote></details>

ah! I see, so another dumb question, if I have not specified any accuracy for fixed step integrator (GetCurrentWitnessTimeIsolation returns emtpy), then the witness triggers at the end of the fixed step? Can you point me to where this behavior is documented? I haven't been following the previous pr very closely. It might be a good idea to add a one liner doc somewhere here. 

---

*[drake/systems/analysis/test/simulator_test.cc, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0QHQK8N7gavo88VIp:-Km5rT9OiDo1BXpaoiAc:bamt78y) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L252)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I changed `w1` to `clock_eval` and `w2` to `logistic_eval`. What did you want me to change instead?
</blockquote></details>

oh! sorry. 

---

*[drake/systems/analysis/test/simulator_test.cc, line 331 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6K6-oyP5O3RKDrTLE:-Km6K6-oyP5O3RKDrTLF:babua71) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L331)):*
> ```cpp
>   const double dt = 2;
>   simulator = std::make_unique<Simulator<double>>(system);
>   simulator->set_publish_at_initialization(false);
> ```

BTW you can call DisableDefaultPublishing

---

*[drake/systems/analysis/test/simulator_test.cc, line 346 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6LRWpCgA7exH3gtGh:-Km6LRWpCgA7exH3gtGi:b63v1f3) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L346)):*
> ```cpp
> }
> 
> // Tests ability of simulation to identify two witness functions triggering
> ```

I don't think this description really matches what you are testing. 

---

*[drake/systems/analysis/test/simulator_test.cc, line 364 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6L7xqH2n8hBy8jgj0:-Km6L7xqH2n8hBy8jgj1:btm477o) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L364)):*
> ```cpp
>   const double dt = 3;
>   Simulator<double> simulator(system);
>   simulator.set_publish_at_initialization(false);
> ```

BTW can call DisableDefaultPublish

---

*[drake/systems/analysis/test/simulator_test.cc, line 1185 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6LYk1Snm5WhnPyOi_:-Km6LYk1Snm5WhnPyOia:b-47guht) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L1185)):*
> ```cpp
> }  // namespace drake
> 
> int main(int argc, char **argv) {
> ```

BTW, Why do you need this? 

---

*[drake/systems/analysis/test/simulator_test.cc, line 121 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km5seKdQNTF7fhgwYFE:-Km5seKdQNTF7fhgwYFF:b-703850) ([raw file](https://github.com/robotlocomotion/drake/blob/53f36a27aa51d00d9e5e721804c783a1a928451d/drake/systems/analysis/test/simulator_test.cc#L121)):*
> ```cpp
> }
> 
> // Initializes the Simulator's integrator to fixed step mode for witness
> ```

probably need to update this.

---

*[drake/systems/analysis/test/simulator_test.cc, line 126 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km5sfpzcUTAtkEPvNOW:-Km5sfpzcUTAtkEPvNOX:belxtjt) ([raw file](https://github.com/robotlocomotion/drake/blob/53f36a27aa51d00d9e5e721804c783a1a928451d/drake/systems/analysis/test/simulator_test.cc#L126)):*
> ```cpp
>   const System<double>& system = s->get_system();
>   auto context = s->get_mutable_context();
>   s->reset_integrator<RungeKutta3Integrator<double>>(system, context);
> ```

BTW, excuse my ignorance, i thought rk3 is fixed step? 
It seems the fixed step or not is controlled by the mode setting method, and RK2 doesn't support error estimate and thus not mode settable with set_fixed_step_mode. So you are using RK3 here, and RK2 above? Is there any reason why not to use RK3 for all these methods? 

---

*[drake/systems/analysis/test/simulator_test.cc, line 206 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6IQPHmPKVSAeO6FdG:-Km6IQPHmPKVSAeO6FdH:b2yphiy) ([raw file](https://github.com/robotlocomotion/drake/blob/53f36a27aa51d00d9e5e721804c783a1a928451d/drake/systems/analysis/test/simulator_test.cc#L206)):*
> ```cpp
>     // not have total control over the endpoints of the continuous interval
>     // (as we do for the FixedStepIncreasingIsolationAccuracy test, immediately
>     // below). Correspondingly, accuracy has to be significantly tightened to
> ```

Can you elaborate more on the correspondingly part? It's not immediately obvious to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
One note: this PR addresses issue #6184

---

Review status: 2 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/analysis/test/simulator_test.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km0DWl2c8ryYMNDZTVy:-Km7O8S75UmW0bAA81tt:b-qybtr1) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L124)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

ah! I see, so another dumb question, if I have not specified any accuracy for fixed step integrator (GetCurrentWitnessTimeIsolation returns emtpy), then the witness triggers at the end of the fixed step? Can you point me to where this behavior is documented? I haven't been following the previous pr very closely. It might be a good idea to add a one liner doc somewhere here. 
</blockquote></details>

Sure- PTAL at the updated documentation here and in `Simulator::GetCurrentWitnessTimeIsolation()`.

---

*[drake/systems/analysis/test/simulator_test.cc, line 346 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6LRWpCgA7exH3gtGh:-Km7RU1YxJPiw5CZ_MPM:bxojn9w) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L346)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

I don't think this description really matches what you are testing. 
</blockquote></details>

It does test what it declares, although you have to be aware of the inner workings of Simulator to understand. PTAL at the updated documentation.

---

*[drake/systems/analysis/test/simulator_test.cc, line 1185 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6LYk1Snm5WhnPyOi_:-Km7Sk4geMRv2-jjeLqO:b93b9tc) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L1185)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, Why do you need this? 
</blockquote></details>

BTW It's useful to be able to turn logging on to get diagnostic information about the witness functions. You can then quickly insert the call to "drake::log()->set_level()` to output such info.

---

*[drake/systems/analysis/test/simulator_test.cc, line 121 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km5seKdQNTF7fhgwYFE:-Km7NQDHf3fJtbvi-Vu5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/53f36a27aa51d00d9e5e721804c783a1a928451d/drake/systems/analysis/test/simulator_test.cc#L121)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

probably need to update this.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 126 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km5sfpzcUTAtkEPvNOW:-Km7NUOAvkv-lF7U_v6R:bvm8eq4) ([raw file](https://github.com/robotlocomotion/drake/blob/53f36a27aa51d00d9e5e721804c783a1a928451d/drake/systems/analysis/test/simulator_test.cc#L126)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, excuse my ignorance, i thought rk3 is fixed step? 
It seems the fixed step or not is controlled by the mode setting method, and RK2 doesn't support error estimate and thus not mode settable with set_fixed_step_mode. So you are using RK3 here, and RK2 above? Is there any reason why not to use RK3 for all these methods? 
</blockquote></details>

BTW RK3 is variable step (but is indeed also capable of operating in fixed step mode), and RK2 is fixed step. There's no reason we couldn't use RK3, or any other integrator for fixed stepping. RK3 in fixed step mode doesn't really buy us anything over RK2 (a slight increase in integration order isn't necessary for these tests to pass).

---

*[drake/systems/analysis/test/simulator_test.cc, line 206 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6IQPHmPKVSAeO6FdG:-KmCrmrYkUQ-snsyUrgd:b-ajzynb) ([raw file](https://github.com/robotlocomotion/drake/blob/53f36a27aa51d00d9e5e721804c783a1a928451d/drake/systems/analysis/test/simulator_test.cc#L206)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

Can you elaborate more on the correspondingly part? It's not immediately obvious to me.
</blockquote></details>

PTAL now- your comment caused me to dig further. It seems that there is no way that I can guarantee that the zero is isolated more accurately, so I did the next best thing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-:-Km7Wxa6ujHzNQ5Udz5c:baqdjmq)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/analysis/test/simulator_test.cc, line 346 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6LRWpCgA7exH3gtGh:-KmR2NlaJkIG-4_JFN5D:b8pb1rf) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L346)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

It does test what it declares, although you have to be aware of the inner workings of Simulator to understand. PTAL at the updated documentation.
</blockquote></details>

Thanks for the added comments! 

---

*[drake/systems/analysis/test/simulator_test.cc, line 1185 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-Km6LYk1Snm5WhnPyOi_:-KmR1WlGOohWYgz6BMwS:b-d871im) ([raw file](https://github.com/robotlocomotion/drake/blob/4bfdbfbb6fff5c878c0b012be627cd062ccc3cf3/drake/systems/analysis/test/simulator_test.cc#L1185)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW It's useful to be able to turn logging on to get diagnostic information about the witness functions. You can then quickly insert the call to "drake::log()->set_level()` to output such info.
</blockquote></details>

good to know, thanks!

---

*[drake/systems/analysis/test/simulator_test.cc, line 165 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmR7VlBnP8HX_jq0xgZ:-KmR7VlBnP8HX_jq0xg_:bs9vd1) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L165)):*
> ```cpp
> // zero to perfect tolerance because the step taken by the integrator
> // fortuitously lands on the zero.
> GTEST_TEST(SimulatorTest, VariableStepIsolation) {
> ```

BTW, this test becomes a different test from the original one. (testing shrinking of the time window used for isolation)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-:-KmR84pixXAvY8-XWS3I:bnfp4nl)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Thanks for the review Siyuan!

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-:-KmRt_AsZ1Fxp6Bkv-xw:bhx9l2q)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
+@jwnimmer-tri for platform review per regular rotation please

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-:-KmRth_Z-uU7J_rchC4a:bd46su8)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+(status: curate commits before merging) 

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending open discussions.

---

Reviewed 1 of 3 files at r1, 1 of 2 files at r2, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/analysis/simulator.h, line 286 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuIgz66hnstS2agzr:-KmRuIgz66hnstS2agzs:b-hibvx2) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/simulator.h#L286)):*
> ```objc
>   /// @returns the isolation window if the Simulator should be isolating
>   ///          witness-triggered events in time, or returns empty otherwise
>   ///          (indicating that any witness-triggered events should trigger
> ```

BTW nit This open-paren is missing a close-paren.

---

*[drake/systems/analysis/test/simulator_test.cc, line 72 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRv1u3aYneg9TQsY2o:-KmRv1u3aYneg9TQsY2p:b-v72sq) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L72)):*
> ```cpp
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(TwoWitnessEmptySystem)
> 
>   explicit TwoWitnessEmptySystem(const double& off1, const double& off2) {
> ```

FYI Weird to pass double by-reference.

---

*[drake/systems/analysis/test/simulator_test.cc, line 139 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRwarMP9tSx4I1BFmV:-KmRwarMP9tSx4I1BFmW:bimwrhf) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L139)):*
> ```cpp
>   system.set_publish_callback([&](const Context<double>& context) {
>     publish_time = context.get_time();
>   });
> ```

FYI This publish-callback-check-publish-time setup code is repeated throughout this PR.  It's not quite defective yet, but its very close.  I think this PR could make better use of test fixtures to reduce boilerplate.  A 1200-line unit test is a bit overwhelming (a problem beyond just this few-lines repetition).

---

*[drake/systems/analysis/test/simulator_test.cc, line 157 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRwJJu5lR5zQMyd5Wu:-KmRwJJu5lR5zQMyd5Wv:b-gck8ca) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L157)):*
> ```cpp
> 
> // Tests the witness function isolation window gets smaller *when Simulator
> // uses a variable step integrator*  as the accuracy in the context becomes
> ```

BTW nit Mid-sentence double space.

---

*[drake/systems/analysis/test/simulator_test.cc, line 182 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRx6NnpCzSBFcPrvHV:-KmRx6NnpCzSBFcPrvHW:boxuzk4) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L182)):*
> ```cpp
>   context->set_accuracy(accuracy);
> 
>   // Set the initial logistic system evaluation.
> ```

FYI Is "logistic system" what you want here?  The DUT appears to be an EmptySystem.

Same for the subsequent test case, as well.

---

*[drake/systems/analysis/test/simulator_test.cc, line 307 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRya8uLay0KxinEa8J:-KmRya8uLay0KxinEa8K:bskvkl1) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L307)):*
> ```cpp
> 
>   // We expect exactly two triggerings.
>   EXPECT_EQ(triggers.size(), 2);
> ```

FYI `ASSERT_EQ` would be better, since the subsequent code might segfault if this doesn't pass.

---

*[drake/systems/analysis/test/simulator_test.cc, line 311 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRyXhbn8_D5W3XeZOX:-KmRyXhcHcbmwS_CrFVD:b-rcwa05) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L311)):*
> ```cpp
>   // Check that the witnesses triggered in the order we expect.
>   EXPECT_TRUE(dynamic_cast<const LogisticWitness*>(triggers.front().second));
>   EXPECT_TRUE(dynamic_cast<const ClockWitness*>(triggers.back().second));
> ```

FYI Prefer https://github.com/RobotLocomotion/drake/blob/master/drake/common/test/is_dynamic_castable.h for these.

---

*[drake/systems/analysis/test/simulator_test.cc, line 408 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRyxH56fN2v7tYvpfO:-KmRyxH56fN2v7tYvpfP:bqaqjbf) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L408)):*
> ```cpp
>   // Verify that the publishes are at the expected times.
>   // NOTE: value_or(999) necessary to work around Mac OS X bug where value()
>   // function is declared but not defined.
> ```

BTW Not necessary anymore.

---

*[drake/systems/analysis/test/simulator_test.cc, line 1197 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuZqDr_4HVno9taSl:-KmRuZqDr_4HVno9taSm:bg3o23b) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L1197)):*
> ```cpp
>   ::testing::InitGoogleTest(&argc, argv);
>   return RUN_ALL_TESTS();
> }
> ```

Given the current `BUILD` rules, `main` should not be included here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-:-KmRu6m6Ha509odjxw0J:b-njh8n5)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: 1 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/analysis/simulator.h, line 286 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuIgz66hnstS2agzr:-KmS-LoqZCjmWa9AkfCw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/simulator.h#L286)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW nit This open-paren is missing a close-paren.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 72 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRv1u3aYneg9TQsY2o:-KmS-P3ZiS5HsWcRGYD_:bf6pa6b) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L72)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Weird to pass double by-reference.
</blockquote></details>

Done (weird- I don't recall doing this).

---

*[drake/systems/analysis/test/simulator_test.cc, line 139 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRwarMP9tSx4I1BFmV:-KmS-p0pjr_FlHmOFEiQ:b-rly07l) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L139)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI This publish-callback-check-publish-time setup code is repeated throughout this PR.  It's not quite defective yet, but its very close.  I think this PR could make better use of test fixtures to reduce boilerplate.  A 1200-line unit test is a bit overwhelming (a problem beyond just this few-lines repetition).
</blockquote></details>

FYI I put in a TODO to remind me to streamline this beast.

---

*[drake/systems/analysis/test/simulator_test.cc, line 157 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRwJJu5lR5zQMyd5Wu:-KmS043RdWNartjB6RjQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L157)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW nit Mid-sentence double space.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 182 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRx6NnpCzSBFcPrvHV:-KmS0CTeBfaayOvbtd7o:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L182)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Is "logistic system" what you want here?  The DUT appears to be an EmptySystem.

Same for the subsequent test case, as well.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 307 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRya8uLay0KxinEa8J:-KmS0FV4RX7oPNW7OOiY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L307)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI `ASSERT_EQ` would be better, since the subsequent code might segfault if this doesn't pass.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 311 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRyXhbn8_D5W3XeZOX:-KmSAyY3QLflKEhpF7gV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L311)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Prefer https://github.com/RobotLocomotion/drake/blob/master/drake/common/test/is_dynamic_castable.h for these.
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 408 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRyxH56fN2v7tYvpfO:-KmS0_GnSkVy4t4nzWKi:b-enhepb) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L408)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Not necessary anymore.
</blockquote></details>

BTW Sweet!

---

*[drake/systems/analysis/test/simulator_test.cc, line 1197 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuZqDr_4HVno9taSl:-KmS0cxsD7Gm4h6L59sJ:bdcd6e4) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L1197)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Given the current `BUILD` rules, `main` should not be included here.
</blockquote></details>

Could you elaborate? I'm not aware of the issue you're broaching.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 1197 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuZqDr_4HVno9taSl:-KmSBcNokXRwhphW0Wyu:b2dkmcr) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L1197)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Could you elaborate? I'm not aware of the issue you're broaching.
</blockquote></details>

This PR adds a `main` function; by default (and indeed in this directory's BUILD file), `drake_cc_googletest` already provides a `main` function that does exactly this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 1197 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuZqDr_4HVno9taSl:-KmSDCVosP2pQ8y4f033:b-5ipbwe) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L1197)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This PR adds a `main` function; by default (and indeed in this directory's BUILD file), `drake_cc_googletest` already provides a `main` function that does exactly this.
</blockquote></details>

Ok- I like to keep the main() function around because it allows me to easily add logging, which is quite useful for debugging witness function isolation issues. Given that goal, what do you suggest?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 1197 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuZqDr_4HVno9taSl:-KmSKOSu3gJnVeqjGxta:b-sr00o7) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L1197)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Ok- I like to keep the main() function around because it allows me to easily add logging, which is quite useful for debugging witness function isolation issues. Given that goal, what do you suggest?
</blockquote></details>

Change the if-else block at https://github.com/RobotLocomotion/drake/blob/da7563bac95864cf103ce7ca3469af59323696e7/tools/drake.bzl#L198-L201 to instead read something like:
```
deps.append("@gtest//:without_main")
if use_default_main:
    deps.append("//tools:gtest_default_main")
```
and then add `drake_cc_library(name = 'gtest_default_main' ,. ..)` with a `cc` file that provides the `main` routine you want, which would be this content _plus_ the use of https://github.com/RobotLocomotion/drake/blob/master/drake/common/text_logging_gflags.h to allow the user to control logging from the command line.

Note that to write this `main` function, you need to do it in the right order, like https://codeyarns.com/2016/02/05/how-to-mix-gflags-and-gtest/ IIRC.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 1197 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuZqDr_4HVno9taSl:-KmTBW3sNHh9tL1iggM8:bnaryu3) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L1197)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Change the if-else block at https://github.com/RobotLocomotion/drake/blob/da7563bac95864cf103ce7ca3469af59323696e7/tools/drake.bzl#L198-L201 to instead read something like:
```
deps.append("@gtest//:without_main")
if use_default_main:
    deps.append("//tools:gtest_default_main")
```
and then add `drake_cc_library(name = 'gtest_default_main' ,. ..)` with a `cc` file that provides the `main` routine you want, which would be this content _plus_ the use of https://github.com/RobotLocomotion/drake/blob/master/drake/common/text_logging_gflags.h to allow the user to control logging from the command line.

Note that to write this `main` function, you need to do it in the right order, like https://codeyarns.com/2016/02/05/how-to-mix-gflags-and-gtest/ IIRC.
</blockquote></details>

I did the latter part (that doesn't require editing drake.bzl). I still don't understand why drake.bzl would need to be modified- it (seemingly?) works fine as-is, and I don't understand the rationale behind the changes that you are proposing, or what the proposed changes seek to do. Would you mind explaining further?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 1197 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuZqDr_4HVno9taSl:-KmTK6ovDYBq3wCrHw1Z:bp69418) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L1197)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I did the latter part (that doesn't require editing drake.bzl). I still don't understand why drake.bzl would need to be modified- it (seemingly?) works fine as-is, and I don't understand the rationale behind the changes that you are proposing, or what the proposed changes seek to do. Would you mind explaining further?
</blockquote></details>

Sure.

Right now, `simulator_test.cc` provides an `main` function in its source code (and object code).  Its `BUILD` rule is `drake_cc_googletest(name = "simulator_test", ...)` ends up using `use_default_main = 1` from `drake.bzl` which does `deps.append("@gtest//:main")` under the hood.  The library `@gtest//:main` object code _also_ provides a `main` function.  In general, we should _not_ have two objects try to provide a `main` function, because its difficult to reason about which `main` function is going to end up being used.  So in short, the defect is that two object files try to provide `main` into the same binary.

The way I was suggesting fixing it was to remove the `main` function from this file in particular, and instead change the default main function used by `drake_cc_googletest` to have verbosity settings that you can then employ on _any_ unit test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/analysis/test/simulator_test.cc, line 1197 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuZqDr_4HVno9taSl:-KmXACsmYQjkvTAXHcH7:b-4jfu5q) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L1197)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Sure.

Right now, `simulator_test.cc` provides an `main` function in its source code (and object code).  Its `BUILD` rule is `drake_cc_googletest(name = "simulator_test", ...)` ends up using `use_default_main = 1` from `drake.bzl` which does `deps.append("@gtest//:main")` under the hood.  The library `@gtest//:main` object code _also_ provides a `main` function.  In general, we should _not_ have two objects try to provide a `main` function, because its difficult to reason about which `main` function is going to end up being used.  So in short, the defect is that two object files try to provide `main` into the same binary.

The way I was suggesting fixing it was to remove the `main` function from this file in particular, and instead change the default main function used by `drake_cc_googletest` to have verbosity settings that you can then employ on _any_ unit test.
</blockquote></details>

... and of course, amending `drake_cc_googletest` could be a separate PR is you wish; just remove the new `main` function here and then this PR would be ready to go.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/analysis/test/simulator_test.cc, line 1197 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmRuZqDr_4HVno9taSl:-KmXFoJq7ZTcPmJmAH2N:bdz8utd) ([raw file](https://github.com/robotlocomotion/drake/blob/5948750f8e87625ed97ed01aed1246440e8a3e53/drake/systems/analysis/test/simulator_test.cc#L1197)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

... and of course, amending `drake_cc_googletest` could be a separate PR is you wish; just remove the new `main` function here and then this PR would be ready to go.
</blockquote></details>

Thanks Jeremy- I think that's probably the right approach. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 2 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/analysis/BUILD, line 170 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmXKk6HDwZLUAXOjiB7:-KmXKk6HDwZLUAXOjiB8:b-ciom1o) ([raw file](https://github.com/robotlocomotion/drake/blob/c681901925bcfc4f198433f55de18a34beef2211/drake/systems/analysis/BUILD#L170)):*
> ```Python
>         ":simulator",
>         "//drake/common:is_dynamic_castable",
>         "//drake/common:text_logging_gflags",
> ```

Unused?

---

*[drake/systems/analysis/BUILD, line 172 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmXKkdKH65ZBfWK-vIx:-KmXKkdKH65ZBfWK-vIy:b-ciom1o) ([raw file](https://github.com/robotlocomotion/drake/blob/c681901925bcfc4f198433f55de18a34beef2211/drake/systems/analysis/BUILD#L172)):*
> ```Python
>         "//drake/common:text_logging_gflags",
>         "//drake/systems/analysis/test/controlled_spring_mass_system",
>         "@com_github_gflags_gflags//:gflags",
> ```

Unused?

---

*[drake/systems/analysis/test/simulator_test.cc, line 9 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmXKdJaeuG0MI2tdg94:-KmXKdJaeuG0MI2tdg95:b-ciom1o) ([raw file](https://github.com/robotlocomotion/drake/blob/c681901925bcfc4f198433f55de18a34beef2211/drake/systems/analysis/test/simulator_test.cc#L9)):*
> ```cpp
> 
> #include <gtest/gtest.h>
> #include <gflags/gflags.h>
> ```

Unused?

---

*[drake/systems/analysis/test/simulator_test.cc, line 16 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmXKe1zFlv-1P6TM5-M:-KmXKe1zFlv-1P6TM5-N:b-ciom1o) ([raw file](https://github.com/robotlocomotion/drake/blob/c681901925bcfc4f198433f55de18a34beef2211/drake/systems/analysis/test/simulator_test.cc#L16)):*
> ```cpp
> #include "drake/common/drake_copyable.h"
> #include "drake/common/text_logging.h"
> #include "drake/common/text_logging_gflags.h"
> ```

Unused?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/analysis/BUILD, line 170 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmXKk6HDwZLUAXOjiB7:-KmXOZqPGHpZNE1uX90g:b-8jp1qu) ([raw file](https://github.com/robotlocomotion/drake/blob/c681901925bcfc4f198433f55de18a34beef2211/drake/systems/analysis/BUILD#L170)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unused?
</blockquote></details>

You're too good, Jeremy!

---

*[drake/systems/analysis/BUILD, line 172 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmXKkdKH65ZBfWK-vIx:-KmXOb9YIwe4K7EsZhm0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c681901925bcfc4f198433f55de18a34beef2211/drake/systems/analysis/BUILD#L172)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unused?
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 9 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmXKdJaeuG0MI2tdg94:-KmXObo1nTqu-3EhA10f:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c681901925bcfc4f198433f55de18a34beef2211/drake/systems/analysis/test/simulator_test.cc#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unused?
</blockquote></details>

Done.

---

*[drake/systems/analysis/test/simulator_test.cc, line 16 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-KmXKe1zFlv-1P6TM5-M:-KmXOcCmPyNVUA-h-27D:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c681901925bcfc4f198433f55de18a34beef2211/drake/systems/analysis/test/simulator_test.cc#L16)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unused?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
(In this case, we can thank reviewable for showing me exactly the right diffs, which made the stray lines stand out.)

---

Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6255#-:-KmXPLgy_Y_kPI7k2Sek:bn1bli1)*
<!-- Sent from Reviewable.io -->
